votingpool/input_selection.go

```diff
@@ -0,0 +1,286 @@
+	if err != nil {
+		return nil, newError(ErrInputSelection, "failed to get unspent outputs", err)
+	}
+	addrMap, err := groupCreditsByAddr(unspents, p.manager.ChainParams())
+	if err != nil {
+		return nil, err
+	}
+	var inputs []Credit
+	address := startAddress
+	for {
+		log.Debugf("Looking for eligible inputs at address %v", address.addrIdentifier())
+		if candidates, ok := addrMap[address.addr.EncodeAddress()]; ok {
+			var eligibles []Credit
+			for _, c := range candidates {
+				if p.isCreditEligible(c, minConf, chainHeight, dustThreshold) {
+					eligibles = append(eligibles, newCredit(c, address))
+				}
+			}
+			// Make sure the eligibles are correctly sorted.
+			sort.Sort(byAddress(eligibles))
+			inputs = append(inputs, eligibles...)
+		}
+		nAddr, err := nextAddr(p, address.seriesID, address.branch, address.index, lastSeriesID+1)
+		if err != nil {
+			return nil, newError(ErrInputSelection, "failed to get next withdrawal address", err)
+		} else if nAddr == nil {
+			log.Debugf("getEligibleInputs: reached last addr, stopping")
+			break
+		}
+		address = *nAddr
+	}
+	return inputs, nil
+}
+
+// nextAddr returns the next WithdrawalAddress according to the input selection
+// rules: http://opentransactions.org/wiki/index.php/Input_Selection_Algorithm_(voting_pools)
+// It returns nil if the new address' seriesID is >= stopSeriesID.
+func nextAddr(p *Pool, seriesID uint32, branch Branch, index Index, stopSeriesID uint32) (
+	*WithdrawalAddress, error) {
+	branch++
+	series := p.GetSeries(seriesID)
+	if series == nil {
+		return nil, newError(ErrSeriesNotExists, fmt.Sprintf("unknown seriesID: %d", seriesID), nil)
+	}
+	if int(branch) > len(series.publicKeys) {
+		highestIdx, err := p.highestUsedSeriesIndex(seriesID)
+		if err != nil {
+			return nil, err
+		}
+		if index > highestIdx {
+			seriesID++
+			log.Debugf("nextAddr(): reached last branch (%d) and highest used index (%d), "+
+				"moving on to next series (%d)", branch, index, seriesID)
+			index = 0
+		} else {
+			index++
+		}
+		branch = 0
+	}
+
+	if seriesID >= stopSeriesID {
+		return nil, nil
+	}
+
+	addr, err := p.WithdrawalAddress(seriesID, branch, index)
+	if err != nil && err.(Error).ErrorCode == ErrWithdrawFromUnusedAddr {
+		// The used indices will vary between branches so sometimes we'll try to
+		// get a WithdrawalAddress that hasn't been used before, and in such
+		// cases we just need to move on to the next one.
+		log.Debugf("nextAddr(): skipping addr (series #%d, branch #%d, index #%d) as it hasn't "+
+			"been used before", seriesID, branch, index)
+		return nextAddr(p, seriesID, branch, index, stopSeriesID)
+	}
+	return addr, err
+}
+
+// highestUsedSeriesIndex returns the highest index among all of this Pool's
+// used addresses for the given seriesID. It returns 0 if there are no used
+// addresses with the given seriesID.
+func (p *Pool) highestUsedSeriesIndex(seriesID uint32) (Index, error) {
+	maxIdx := Index(0)
+	series := p.GetSeries(seriesID)
```

Comment:
I've noticed this throughout the code, but curious is there is a specific reason, such as a name collision, for why it's named `GetSeries` instead of `Series` since the former is against Effective Go style guidelines?